icu4c/source/common/uresdata.cpp

```diff
@@ -817,10 +817,41 @@ res_getResource(const ResourceData *pResData, const char *key) {
 }
 
 
+icu::ResourceTable::ResourceTable(const uint16_t *k16, const int32_t *k32,
+                                  const uint16_t *i16, const uint32_t *i32, int32_t len,
+                                  const ResourceTracer& traceInfo, const ResourceData *tableData) :
+        keys16(k16), keys32(k32), items16(i16), items32(i32), length(len),
+        fTraceInfo(traceInfo), tableData(new ResourceData(*tableData)) {
+}
+
+icu::ResourceTable::ResourceTable(const ResourceTable& that) :
+        keys16(that.keys16), keys32(that.keys32), items16(that.items16), items32(that.items32), length(that.length),
+        fTraceInfo(that.fTraceInfo), tableData(new ResourceData(*that.tableData)) {
+}
+
+icu::ResourceTable::~ResourceTable() {
+    delete tableData;
+}
+
+icu::ResourceTable& icu::ResourceTable::operator=(const ResourceTable& that) {
+    delete this->tableData;
+
+    this->keys16 = that.keys16;
+    this->keys32 = that.keys32;
+    this->items16 = that.items16;
+    this->items32 = that.items32;
+    this->length = that.length;
+    this->fTraceInfo = that.fTraceInfo;
+    this->tableData = new ResourceData(*that.tableData);
+
+    return *this;
+}
+
 UBool icu::ResourceTable::getKeyAndValue(int32_t i,
                                          const char *&key, icu::ResourceValue &value) const {
     if(0 <= i && i < length) {
         icu::ResourceDataValue &rdValue = static_cast<icu::ResourceDataValue &>(value);
+        rdValue.setData(tableData); // `value` is an output-only parameter, but these macros expect it to point to `tableData`
```

Comment:
I think I remember how this works so far: While iterating through resources in a bundle the ResourceData is always the same. So it never changes, and that's how we haven't gotten into trouble with these loops over getKeyAndValue(). (`resData` could be const if we didn't need to be able to default-construct ResourceDataValue.)

The *resource* (`res`) changes (.setResource()), but that's harmless because before the loop-over-table-items started we copied the table's own resource into its ResourceTable and already used it to fetch the pointers into the key and item arrays.

The problem is that an alias can point to a resource in a different bundle/file, and thus needs a different ResourceData pointer when that happens.

It seems like the solution would be for getKeyAndRealValue() to *return a new* ResourceValue, so that it does not overwrite the input value's data pointer. With that we would not need to add this tableData to ResourceTable.

New problem: ResourceValue is a base interface, so we can't return it by value (and we forbid copying). We could take an output reference, a new one separate from the input value, so that we don't clobber that one's resData, but then callers need to know about ResourceDataValue. Maybe that's acceptable until we figure out a better structure? Or, might as well return a ResourceDataValue by value???

Also: If the alias lives in another bundle, then we may also need to think about the lifetime of that bundle. Either we rely on the resource bundle cache, and convince ourselves that that's ok, or we add an abstract bundle pointer into the ResourceDataValue (yuck), probably with some mini interface that lets us ures_close() the bundle (yuck) without knowing all about UResourceBundle at that level.

----
Crazy idea: We might be able to make aliases completely transparent in the ResourceValue API, while avoiding many of these problems. When we get a value, don't immediately follow the alias. Only when we getString(), getTable(), getBinary(), etc. -- then we follow an alias if there is one, before we really do fetch the string/table/binary/whatever. Those functions already take a UErrorCode, and most of them don't have to return a ResourceValue nor a UResourceBundle. And, by following the alias only when actually accessing the value contents, we avoid wasting work when the caller looks at the key and ignores the value.

Now, we still have the problem with getTable() and getArray() that you are trying to solve here: While their lookup works with just pointers into the key and item arrays, for returning their items we will need to either switch the output value's resData or switch to a new ResourceValue.

FYI, schlepping around the ResourceData would be much simpler if we could make https://unicode-org.atlassian.net/browse/ICU-20769 work; in particular, we could carry an opaque pointer in places that don't need to know the details.

We would also still have to think about the lifetime of an alias-hosting bundle, but again maybe the bundle cache solves that.